specification/dartLangSpec.tex

```diff
@@ -3073,20 +3076,32 @@ \subsection{Constructors}
 where $X_j, j \in 1 .. m$ are the formal type parameters of $C$
 and $T_j, j \in 1 .. m$ are specified in the given context.
 We may also omit such a substitution when the given context is
-the instance scope of $C$, where $X_1, \ldots, X_m$ are in scope.
+the class scope of $C$, where $X_1, \ldots, X_m$ are in scope.
 }
 
-\commentary{
+\commentary{%
 A constructor declaration may conflict with static member declarations
 (\ref{classMemberConflicts}).
 }
 
-% In what scope do constructors go? The simple names of named constructors go in the static scope of the class. Unnamed ones go nowhere, but we use the class name to refer to them; the class name could also in the static scope of the class as well to prevent weird errors, or we could ban it explicitly and avoiding duplication. Similarly, the instance scope could contain the constructor names and class name, or we could have special rules to prevent collisions between instance members and constructors or the class.
-
-% The enclosing scope of a generative constructor is the instance scope of the class in which it is declared (but what about redirecting?)
+\commentary{%
+A constructor declaration does not introduce a name into a scope.
+The resolution of an instance creation expression
+(\ref{new})
+to a particular constructor
+relies on the library scope to determine the class
+(possibly via an import prefix),
+and for a constructor denoted by \code{$C$.\id} or \code{$prefix$.$C$.\id},
+it is directly specified that the class $C$ must declare
+a constructor with the name \code{$C$.\id}.
+In other words, the constructor is never determined via a lexical lookup,
+only the class.%
```

Comment:
This feels weird (and wrong) because we don't know whether `Foo.id()` (no `new`) *is* an instance creation expression without first checking whether `Foo.id` is a constructor or a static member. 

Does "instance creation expression" only refer to `new C.id(...)` and `const C.id(...)`, not the expressions without the `new`/`const`? If so, the text here doesn't handle the most common case, so it doesn't add much v alue.
If not, then it's wrong that the expression `C.id(..)` must denote a constructor, it could be a static function.

That is, we can't really say that an instance creation expression without a `new` *must* denote a constructor, because it isn't an instance creation expression if it doesn't denote a constructor. The `must` requirement is tautologically true.
(We don't "do lookup", we just "look up" whether `Foo` has a `Foo.id` constructor.)

Maybe be more elaborate:

> A constructor declaration does not introduce a name into a scope.
> The resolution of an instance creation expression with a \NEW{} (\ref{new}) or \CONST{} (\ref{const})
> to a particular constructor
> relies on the library scope to determine the class
> (possibly via an import prefix),
> and for a constructor denoted by \code{$C$.\id} or \code{$prefix$.$C$.\id},
> it is directly specified that the class $C$ must declare
> a constructor with the name \code{$C$.\id}.
> An invocation of the form \code{$C$.\id(...)} or \code{$prefix}$.$C$.\id(...)} 
> first resolves $C$ or \code{$prefix$.$C$}, and if that resolves to a class,
> it directly checks whether that class declares a constructor with the name \code{$C$.\id}
> (or whether it perhaps has a static member instead).
> In other words, the constructor is never determined via a \emph{lexical} lookup, only the class.%

Reply:
> we don't know whether `Foo.id()` (no `new`) is an instance 
> creation expression without first checking whether `Foo.id` is
> a constructor or a static member.

The omitted `new`/`const` is handled exactly like that: We add a `new` or a `const` keyword to `Foo.id()` when `Foo.id` denotes a constructor, not when it is a static member, cf. https://github.com/dart-lang/language/blob/6df1612f667f7fdd28dd3b66a89e06b75ea98d2a/specification/dartLangSpec.tex#L10947.

In that step, we just look up the class `$C$`, and then we require explicitly that the declaration of `$C$` contains a declaration of a constructor named `$C$` respectively `$C$.\id`.

So we don't need to worry about how we handle an instance creation of the form `Foo.id()` because it's gone after desugaring, and the desugaring itself looks up the class, not the constructor.

I believe that the elaborated form gives information which is already given in other sections, so I added a reference to \ref{functionExpressionInvocation} at the end of this commentary.

Reply:
Desugaring cannot look up something, like $C$, before scopes exist, so it still feels to me like this text skips an important part - the part where we recognize constructors as constructors before adding the `new`.

It might just be the phrasing

Maybe:

> If an expression denotes a constructor, either as \code{$C$.\id} or \code{$prefix$.$C$.\id},
> or as $C$ or \code{$prefix$.$C$} in an invocation context,
> resolution relies on the library scope to determine the class (possibly via an import prefix),
> and then the class declaration is directly checked for whether it has a constructor named \code{$C$.\id}
> or $C$. It is not possible for an identifier to directly refer to a constructor, since the constructor is not
> in any scope used for resolving identifiers.

Reply:
I like that! Now using the new 'Maybe:' text, with a few adjustments.

> Desugaring cannot look up something, like `$C$`, before scopes exist,

It's not important to me that we call this step desugaring, so if we don't want to use that term to denote a transformation that has access to name resolution (or types, or any other non-context-free properties) then we'd just use a different name.

However, we have lots of transformations where source code is changed, and the changes are influenced by information from static analysis. For example, the transformation that defines the semantics of a for-in statement needs the static type of `$e$`, the transformation that transforms `id` to `this.id` needs the interface of the enclosing class, etc.

I know that Johnni is very worried about every piece of syntactic sugar in the specification (and I'm worried as well, whenever it may turn out to be imprecise), but I do believe that we can maintain a reasonably well-defined notion of source code transformations that are able to use information from static analysis. The crucial point is probably that we can't allow those transformations to change the static analysis of non-transformed expressions. With that, I think it's OK to use this sparingly and cautiously.